protoc-gen-openapiv2/internal/genopenapi/template.go

```diff
@@ -898,14 +899,10 @@ func templateToParts(path string, reg *descriptor.Registry, fields []*descriptor
 			buffer += string(char)
 			jsonBuffer += string(char)
 		case ':':
-			// Only the last part may be a verb (":" LITERAL)
-			if depth == 0 {
-				parts = append(parts, buffer)
-				buffer = ":"
-				continue
-			}
-			buffer += string(char)
-			jsonBuffer += string(char)
+			// As soon as we find a ":", everything following is a verb
```

Comment:
Even within `{}`?

Reply:
Fair point... should it be allowed as part of a variable name? The grammar is:

```
// ### Path template syntax
//
//     Template = "/" Segments [ Verb ] ;
//     Segments = Segment { "/" Segment } ;
//     Segment  = "*" | "**" | LITERAL | Variable ;
//     Variable = "{" FieldPath [ "=" Segments ] "}" ;
//     FieldPath = IDENT { "." IDENT } ;
//     Verb     = ":" LITERAL ;
```

Since this PR is making the assertion that a `LITERAL` can contain a `:` (by allowing it after the first `:` which indicates the start of the verb), I suppose we should also allow it in `Segment`, which would mean allowing it inside `{}`. I'll rewrite this and add another test, and check what our other parser does in this case.

Reply:
Thanks for your quick look at this, I really appreciate it. I changed it so we allow colons in variables and added a test to both parsers.

Reply:
Note that `LITERAL` is not allowed to contain `:`, so supporting more than one colon in the verb is supporting paths outside of the [`HttpRule`](https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L242-L244) specification:
> The syntax `LITERAL` matches literal text in the URL path. If the `LITERAL`
>  contains any reserved character, such characters should be percent-encoded
> before the matching.

As I understand it, 'reserved' here refers to [RFC 6570](https://datatracker.ietf.org/doc/html/rfc6570):
```
reserved       =  gen-delims / sub-delims
gen-delims     =  ":" / "/" / "?" / "#" / "[" / "]" / "@"
sub-delims     =  "!" / "$" / "&" / "'" / "(" / ")"
               /  "*" / "+" / "," / ";" / "="
```

However, since this is already supported in the the `HttpRule` parser, I see the point in also supporting it in the OpenAPI generator. I also back the suggestion on using the same parser both places. The question is if the parser should be strict or not.

Does the `HttpRule` parser also allow other path templates outside the specification? Specifically `:`  before a variable? The reason I am asking is that the the joining of the path in the generator assumes that only the last part can start with `:`.

Reply:
Hm, we were never able to figure out what characters were reserved when making the amendments to the HttpRule parser in https://github.com/grpc-ecosystem/grpc-gateway/pull/1947, where did you find the reference to RFC 6570? It probably makes sense but would be good to have for future reference.

Unfortunately, as you surmise, it's probably out of the question for us to change this behavior now, so we're stuck supporting colons in the verb (and by extension, any LITERAL).

Good question on the `:` preceding a variable, it _probably_ does, I'll add a test and see. In any case, I think that can just be a separate issue from this immediate PR.

Reply:
I've raised #2833 so we don't forget the point you made. I think this PR is still OK to go in as is.

Reply:
I'm just assuming that it refers to `reserved` in RFC 6570, as it's not explicitly defined in `http.proto`, and it makes explicit references to the RFC when it's _not_ followed. I might be wrong though.

I agree it's OK to merge this PR.